File: server/shared/typescript/validation/abstractions/src/interfaces/i-postal-code-validator.ts
// -----------------------------------------------------------------------
// Copyright (c) DCSV. All rights reserved.
// -----------------------------------------------------------------------

import type { CountryCode } from "@d2/geo-abstractions";
import type { D2Result } from "@d2/result";

/**
 * Mirror of .NET `D2.Shared.Validation.Abstractions.IPostalCodeValidator` —
 * country-aware postal-code validator that returns a normalized form on
 * success.
 *
 * Implementations live in `@d2/validation` backed by the default
 * per-country normalization rules; tests can supply ad-hoc fixtures by
 * implementing this interface directly.
 */
export interface IPostalCodeValidator {
  /**
   * Validates the supplied postal code and returns the normalized
   * representation on success.
   *
   * @param postalCode - The postal code to validate (may be `undefined`,
   *   empty, or whitespace).
   * @param countryCode - Optional country whose postal-code format governs
   *   validation. When `undefined` / omitted, validation fails closed —
   *   there is NO country-agnostic structural fallback. An unknown /
   *   unsupported country also fails closed.
   * @returns An `ok` `D2Result` wrapping the normalized (trimmed and
   *   uppercased) postal code on success; a `validationFailed` `D2Result`
   *   with a per-field `InputError` keyed `"postalCode"` carrying
   *   `POSTAL_CODE_INVALID` on `undefined`, empty, whitespace, structurally
   *   invalid, omitted-country, or unknown-country input.
   */
  validate(
    postalCode: string | undefined,
    countryCode?: CountryCode,
  ): D2Result<string>;
}

File: public/packages/dotnet/location/core/IPostalCodeValidator.cs
// -----------------------------------------------------------------------
// <copyright file="IPostalCodeValidator.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Location;

using DcsvIo.D2.Geo.Abstractions;
using DcsvIo.D2.Result;

/// <summary>
/// Boundary validator for postal codes invoked by
/// <see cref="ValueObjects.AdminLocation"/> (and any other consumer
/// that needs structured postal-code validation). Lives in
/// <c>DcsvIo.D2.Location</c> (NOT Abstractions) so the DI seam stays
/// out of pure-vocabulary projects per the Plan §4.5 layer policy.
/// </summary>
public interface IPostalCodeValidator
{
    /// <summary>
    /// Validates a postal code. <paramref name="countryCode"/> is
    /// optional — the default implementation ignores it (global-range
    /// regex only); consumer implementations may use it for
    /// country-specific validation.
    /// </summary>
    /// <param name="postalCode">The postal code to validate (may be null / whitespace).</param>
    /// <param name="countryCode">
    /// Optional ISO 3166-1 alpha-2 country code for country-aware validation.
    /// </param>
    /// <returns>
    /// <c>Ok</c> wrapping the normalized postal code on success;
    /// <see cref="D2Result{TData}.ValidationFailed"/> for null / empty /
    /// out-of-range / out-of-shape inputs.
    /// </returns>
    D2Result<string> Validate(string? postalCode, CountryCode? countryCode = null);
}
